### ISSUE TITLE
Proposal: Add an incremental syntax to CSS Gradients

### ISSUE THREAD

**Issue by Afif13** [w3c/csswg-drafts#8616]
I think it's good to have a kind of "relative values" to express a relation between two consecutive color stops.

Let's say I have `linear-gradient(red 20%,blue 40%,green 60%)` where each time I am adding `20%` to the color stops. It would be good if I can, instead, write `linear-gradient(red 20%,blue +20%,green +20%)`. The plus sign means "take the last color stop and increment it". This way we can easily express the `20%` as a variable and reuse it.

Here are a few more examples:

```css
/* Instead of */
background: linear-gradient(red 50%, blue calc(50% + 1px));
/* We write */
background: linear-gradient(red 50%, blue +1px);

/* Instead of */
background: linear-gradient(red 20%, blue calc(20% + 20px), green calc(20% + 40px));
/* We write */
background: linear-gradient(red 20%, blue +20px, green +20px);

/* Instead of */
background: conic-gradient(red 60deg, blue 120deg, green 180deg, yellow 240deg);
/* We write */
background: conic-gradient(red 60deg, blue +60deg, green +60deg, yellow +60deg);
/* We can also write */
background: conic-gradient(red +60deg, blue +60deg, green +60deg, yellow +60deg);
/* the first +60deg will be relative to 0deg so the same as 60deg */


/* Instead of */
background: repeating-linear-gradient(red 0 10px, blue 0 20px);
/* we write */
background: repeating-linear-gradient(red 0 10px, blue 0 +10px);
/* the browser will transform the 0 after blue to 10px then will use that value as a reference for the next one to have 10px+10px = 20px
/* we can also write it like below */
background: repeating-linear-gradient(red 0 +10px, blue 0 +10px);
/* the "0 +10px" can be a variable */

```

I think the `+` won't create any confusion in the syntax. It can be an optional character before the value to indicate if the value is static or relative.

The calculation should also be done after the [Color stop "Fixup"](https://w3c.github.io/csswg-drafts/css-images-4/#color-stop-fixup) like I did in the last example. When doing `linear-gradient(red 150px, blue 70px +50px)` we should have `150px + 50px` and not `70px + 50px`.

**Comment by romainmenke** [comment]
`+60deg` is just `60deg` with an explicit sign.
But `+ 60deg` with a space between `+` and a dimension token could be a way to express incremental values.

**Comment by Afif13** [comment]
> `+60deg` is just `60deg` with an explicit sign. But `+ 60deg` with a space between `+` and a dimension token could be a way to express incremental values.
> 
> Edit : `++60deg` might be more readable. As CSS tokens this is a delim token `+` and a dimension token with explicit sign.

Yes, but I thought that since the `+` is rarely used to indicate positive values we can do an exception inside gradients and use it as increment. Maybe we should consider another character to avoid conflict?

**Comment by romainmenke** [comment]
`++` has the benefit of being well established in other languages and the intent is the same as it would be here : an implicit read and increment with whatever is on the right of `++`.

The only awkward thing about a syntax like this is that there really isn't a meaningful difference between `++60deg`, `+-60deg` `--60deg`, `-+60deg`.

`++` and `--` aren't real operators in this syntax. One symbol is a delim and the other is the sign of the dimension token.

----

If this is something you want to prototype as a PostCSS plugin, please let me know, happy to help out. I really like the work you do with CSS gradients!

**Comment by Afif13** [comment]
I won't say no to a prototype 😊 if it's something easy to do why not. It can help showing real examples and make the proposal even better! Thank you :+1:

**Comment by brandonmcconnell** [comment]
@Afif13 To make the syntax clear, are you thinking it would always either be a **double**-plus/minus, so either `++60px` or `--60px` for example?

**Comment by Afif13** [comment]
@brandonmcconnell the idea is not using `++` as the operator. As explained by @romainmenke, one plus will the positive sign while the other is the incremental operator so in your last example we would use

```css
background: linear-gradient(red 20%, blue +var(--default-shift), green +calc(var(--default-shift) * 2));
```

Same logic with `--` but I won't focus on this part because incrementing is the only useful operation inside gradient since we always need to have bigger values. Smaller values are not useful that's why my proposal is only about "incremental syntax".

It can also be another character like `~` to avoid confusion but in all the cases, the idea is to have an optional character to indicate if it's a static value or a relative one.

The `shift()` function is good but .. a bit verbose 😅 I prefer a reduced syntax but if a function is the only way, I would take it.

**Comment by cdoublev** [comment]
You might be interested in taking inspiration from the [relative color syntax](https://drafts.csswg.org/css-color-5/#relative-colors), since the idea seems to express relative (to the previous color stop) vs. absolute color stop positions.

For example, `linear-gradient(red, orange calc(prev-stop + 50%), green)`.

**Comment by brandonmcconnell** [comment]
@Afif13 Yeah, either syntax would be good in my opinion— `+` or `shift()`.

Also, I just did some testing, and it looks like either `++50%` or `+(50%)` could be acceptable since neither are currently valid.

I thought `+(50%)` might be valid syntax already, but it appears to break the style as far as I can see. I'm not suggesting we strictly stick to one of these but rather than either of these could be acceptable uses.

This is pretty similar to how we can use properties on numbers in JS, for example:

```js
(5).toFixed(2)
5..toFixed(2)
```

Either of the above are valid, but `5.toFixed(2)` is not since a single dot without parentheses could be mistaken for a decimal delimiter.

Similarly in your proposal, a single sign could be valid outside parentheses, or two symbols adjacent without parentheses, which could be validly any combination:

* `++5px`
* `+-5px`
* `-+5px`
* `--5px`

And they each do different things, where the first sign dictates whether to add or remove the amount, and the second is the polarity of the actual number.

**Comment by brandonmcconnell** [comment]
For anyone interested, I ported @romainmenke's plugin to CodePen for quicker tinkering, here:

https://codepen.io/brandonmcconnell/pen/YzOOBVj/66c65611ba8bf3133f1cf87ccfe09c44?editors=0110

Really enjoying this syntax, @Afif13. Great suggestion 💯

**Comment by romainmenke** [comment]
How does this work when color stop positions are omitted?

Does first fill in the color stops as if the relative value was also missing?
And then afterwards combine the interpolated position from the previous color stop with the relative value?

Or does it first apply the relative value as if the previous color stop had a value of zero?

---------

```
linear-gradient(red, white ++10%, blue);
```

Is that :

- `linear-gradient(red 0%, white 10%, blue 100%);`
- `linear-gradient(red 0%, white 60%, blue 100%);` -> definitely no

---------

```
linear-gradient(red, orange, white, pink ++10%, blue);
```

Is that :

- `linear-gradient(red 0%, orange 3.3333%, white 6.6666%, pink 10%, blue 100%);`
- `linear-gradient(red 0%, orange 25%, white 50%, pink 60%, blue 100%);`
- `linear-gradient(red 0%, orange 25%, white 50%, pink 85%, blue 100%);` -> definitely no

---------

```
linear-gradient(red, orange ++10%, white, pink, blue);
```

Is that :

- `linear-gradient(red 0%, orange 10%, white 40%, pink 70%, blue 100%);`
- `linear-gradient(red 0%, orange 10%, white 50%, pink 75%, blue 100%);`
- `linear-gradient(red 0%, orange 35%, white 50%, pink 75%, blue 100%);` -> definitely no

**Comment by Afif13** [comment]
@romainmenke 

Actually, I never thought about this but now I can see a lot of complex cases to consider 😅

The issue I can see is an example like `linear-gradient(red, blue ++50%, pink, green)`. If we first calculate all the color stops (including the pink one) we can easily find ourselves with the final value of `++50%` bigger that the value of `pink` and we will need another Fixup step which is not good.

I think the suitable solution is to consider a relative color as static if there is no reference for it (no color stop is defined before it). The previous example will become `linear-gradient(red, blue 50%, pink, green)` then we do the color Fixup to get `linear-gradient(red 0%, blue 50%, pink 75%, green 100%)`

So like you said "it first apply the relative value as if the previous color stop had a value of zero" sounds good to me.

**Comment by Afif13** [comment]
pinging the Spec Editors so they can give us their feedback 🙏

@fantasai @LeaVerou @tabatkins
